File: backend/app/services/s3.py
import boto3
from botocore.config import Config as BotoConfig
from app.config import get_settings


def get_s3_client():
    settings = get_settings()
    return boto3.client(
        "s3",
        aws_access_key_id=settings.AWS_ACCESS_KEY_ID,
        aws_secret_access_key=settings.AWS_SECRET_ACCESS_KEY,
        region_name=settings.AWS_REGION,
        config=BotoConfig(signature_version="s3v4"),
    )


def generate_presigned_upload_url(key: str, content_type: str, expires_in: int = 3600) -> str:
    settings = get_settings()
    client = get_s3_client()
    return client.generate_presigned_url(
        "put_object",
        Params={
            "Bucket": settings.S3_BUCKET_NAME,
            "Key": key,
            "ContentType": content_type,
        },
        ExpiresIn=expires_in,
    )


def generate_presigned_download_url(key: str, expires_in: int = 900) -> str:
    settings = get_settings()
    client = get_s3_client()
    return client.generate_presigned_url(
        "get_object",
        Params={"Bucket": settings.S3_BUCKET_NAME, "Key": key},
        ExpiresIn=expires_in,
    )


def upload_file_to_s3(key: str, file_bytes: bytes, content_type: str) -> None:
    settings = get_settings()
    client = get_s3_client()
    client.put_object(
        Bucket=settings.S3_BUCKET_NAME,
        Key=key,
        Body=file_bytes,
        ContentType=content_type,
    )


def download_file_from_s3(key: str) -> bytes:
    settings = get_settings()
    client = get_s3_client()
    response = client.get_object(Bucket=settings.S3_BUCKET_NAME, Key=key)
    return response["Body"].read()


def delete_file_from_s3(key: str) -> None:
    settings = get_settings()
    client = get_s3_client()
    client.delete_object(Bucket=settings.S3_BUCKET_NAME, Key=key)


def delete_prefix_from_s3(prefix: str) -> None:
    settings = get_settings()
    client = get_s3_client()
    response = client.list_objects_v2(Bucket=settings.S3_BUCKET_NAME, Prefix=prefix)
    if "Contents" in response:
        objects = [{"Key": obj["Key"]} for obj in response["Contents"]]
        client.delete_objects(
            Bucket=settings.S3_BUCKET_NAME,
            Delete={"Objects": objects},
        )

File: src/services/s3.ts
import {
  S3Client,
  PutObjectCommand,
  DeleteObjectCommand,
  DeleteObjectsCommand,
  GetObjectCommand,
  ListObjectsV2Command,
} from '@aws-sdk/client-s3';
import { getSignedUrl } from '@aws-sdk/s3-request-presigner';
import { config } from '../config';
import { Readable } from 'stream';

function getS3Client(): S3Client {
  return new S3Client({
    region: config.AWS_REGION,
    credentials: {
      accessKeyId: config.AWS_ACCESS_KEY_ID,
      secretAccessKey: config.AWS_SECRET_ACCESS_KEY,
    },
    requestChecksumCalculation: 'WHEN_REQUIRED',
    responseChecksumValidation: 'WHEN_REQUIRED',
  });
}

export async function generatePresignedUploadUrl(key: string, contentType: string, expiresIn = 3600): Promise<string> {
  const client = getS3Client();
  const cmd = new PutObjectCommand({ Bucket: config.S3_BUCKET_NAME, Key: key, ContentType: contentType });
  return getSignedUrl(client, cmd, { expiresIn });
}

export async function generatePresignedDownloadUrl(key: string, expiresIn = 900): Promise<string> {
  const client = getS3Client();
  const cmd = new GetObjectCommand({ Bucket: config.S3_BUCKET_NAME, Key: key });
  return getSignedUrl(client, cmd, { expiresIn });
}

export async function uploadFileToS3(key: string, body: Buffer, contentType: string): Promise<void> {
  const client = getS3Client();
  await client.send(new PutObjectCommand({ Bucket: config.S3_BUCKET_NAME, Key: key, Body: body, ContentType: contentType }));
}

export async function deleteFileFromS3(key: string): Promise<void> {
  const client = getS3Client();
  await client.send(new DeleteObjectCommand({ Bucket: config.S3_BUCKET_NAME, Key: key }));
}

export async function deletePrefixFromS3(prefix: string): Promise<void> {
  const client = getS3Client();
  const list = await client.send(new ListObjectsV2Command({ Bucket: config.S3_BUCKET_NAME, Prefix: prefix }));
  if (!list.Contents?.length) return;
  const objects = list.Contents.map((o) => ({ Key: o.Key! }));
  await client.send(new DeleteObjectsCommand({ Bucket: config.S3_BUCKET_NAME, Delete: { Objects: objects } }));
}

export async function getS3ObjectStream(key: string): Promise<Readable> {
  const client = getS3Client();
  const response = await client.send(new GetObjectCommand({ Bucket: config.S3_BUCKET_NAME, Key: key }));
  return response.Body as Readable;
}
